Configure PKI directory and file to correct owner and permissions

floryut opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Configure pki directory to root:root instead of kube:root
Change file permissions to 600 for *.key and 644 for *.crt

**Which issue(s) this PR fixes**:
Fixes #7451

**Special notes for your reviewer**:
None

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

floryut commented:
> Aren't those keys created by kubeadm ? Also any container doing a bind mount of those keys need to run as root if we change those file permissions, not sure this is the right way to go

Well, not sure why a container would need to mount key files, do you have a use case in mind ?

champtar commented:
I mean `apiserver.key` is definitely mounted by the apiserver, but yeah everything run as root :(
```
spec:
  containers:
  - command:
    - kube-apiserver
...
    - --client-ca-file=/etc/kubernetes/ssl/ca.crt
...
    - --etcd-cafile=/etc/ssl/etcd/ssl/ca.pem
    - --etcd-certfile=/etc/ssl/etcd/ssl/node-ech-el8-ks151.pem
    - --etcd-keyfile=/etc/ssl/etcd/ssl/node-ech-el8-ks151-key.pem
...
    - --kubelet-client-certificate=/etc/kubernetes/ssl/apiserver-kubelet-client.crt
    - --kubelet-client-key=/etc/kubernetes/ssl/apiserver-kubelet-client.key
...
    - --proxy-client-cert-file=/etc/kubernetes/ssl/front-proxy-client.crt
    - --proxy-client-key-file=/etc/kubernetes/ssl/front-proxy-client.key
...
    - --requestheader-allowed-names=front-proxy-client
    - --requestheader-client-ca-file=/etc/kubernetes/ssl/front-proxy-ca.crt
    - --tls-cert-file=/etc/kubernetes/ssl/apiserver.crt
    - --tls-private-key-file=/etc/kubernetes/ssl/apiserver.key
    image: k8s.gcr.io/kube-apiserver:v1.18.15
...
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /etc/ssl/certs
      name: ca-certs
      readOnly: true
    - mountPath: /etc/pki
      name: etc-pki
      readOnly: true
    - mountPath: /etc/pki/ca-trust
      name: etc-pki-ca-trust
      readOnly: true
    - mountPath: /etc/pki/tls
      name: etc-pki-tls
      readOnly: true
    - mountPath: /etc/ssl/etcd/ssl
      name: etcd-certs-0
      readOnly: true
    - mountPath: /etc/kubernetes/ssl
      name: k8s-certs
      readOnly: true
```

champtar commented:
```
[root@ech-el8-ks151 ssl]# ls -la
total 52
drwxr-xr-x. 2 kube root  288 Mar 25 00:51 .
drwxr-xr-x. 4 kube root 4096 Mar 25 00:56 ..
-rw-r--r--. 1 root root 1099 Mar 25 00:51 apiserver-kubelet-client.crt
-rw-------. 1 root root 1679 Mar 25 00:51 apiserver-kubelet-client.key
-rw-r--r--. 1 root root 1582 Mar 25 00:51 apiserver.crt
-rw-------. 1 root root 1679 Mar 25 00:51 apiserver.key
-rw-r--r--. 1 root root 1025 Mar 25 00:51 ca.crt
-rw-------. 1 root root 1675 Mar 25 00:51 ca.key
-rw-r--r--. 1 root root 1038 Mar 25 00:51 front-proxy-ca.crt
-rw-------. 1 root root 1679 Mar 25 00:51 front-proxy-ca.key
-rw-r--r--. 1 root root 1058 Mar 25 00:51 front-proxy-client.crt
-rw-------. 1 root root 1679 Mar 25 00:51 front-proxy-client.key
-rw-------. 1 root root 1679 Mar 25 00:51 sa.key
-rw-------. 1 root root  451 Mar 25 00:51 sa.pub
```

floryut commented:
> ```
> [root@ech-el8-ks151 ssl]# ls -la
> total 52
> drwxr-xr-x. 2 kube root  288 Mar 25 00:51 .
> drwxr-xr-x. 4 kube root 4096 Mar 25 00:56 ..
> -rw-r--r--. 1 root root 1099 Mar 25 00:51 apiserver-kubelet-client.crt
> -rw-------. 1 root root 1679 Mar 25 00:51 apiserver-kubelet-client.key
> -rw-r--r--. 1 root root 1582 Mar 25 00:51 apiserver.crt
> -rw-------. 1 root root 1679 Mar 25 00:51 apiserver.key
> -rw-r--r--. 1 root root 1025 Mar 25 00:51 ca.crt
> -rw-------. 1 root root 1675 Mar 25 00:51 ca.key
> -rw-r--r--. 1 root root 1038 Mar 25 00:51 front-proxy-ca.crt
> -rw-------. 1 root root 1679 Mar 25 00:51 front-proxy-ca.key
> -rw-r--r--. 1 root root 1058 Mar 25 00:51 front-proxy-client.crt
> -rw-------. 1 root root 1679 Mar 25 00:51 front-proxy-client.key
> -rw-------. 1 root root 1679 Mar 25 00:51 sa.key
> -rw-------. 1 root root  451 Mar 25 00:51 sa.pub
> ```

Looking at some CI jobs, at least 2 `.crt` file needs mode change, the other are indeed ok (and all `.key` files seems fine)
```
changed: [instance-2] => (item={'uid': 0, 'woth': False, 'mtime': 1617813389.188506, 'inode': 21659241, 'isgid': False, 'size': 1066, 'roth': False, 'isuid': False, 'isreg': True, 'pw_name': 'root', 'gid': 0, 'ischr': False, 'wusr': True, 'xoth': False, 'rusr': True, 'nlink': 1, 'issock': False, 'rgrp': False, 'gr_name': 'root', 'path': '/etc/kubernetes/ssl/ca.crt', 'xusr': False, 'atime': 1617813389.4215074, 'isdir': False, 'ctime': 1617813389.188506, 'isblk': False, 'xgrp': False, 'dev': 64769, 'wgrp': False, 'isfifo': False, 'mode': '0600', 'islnk': False}) => {"ansible_loop_var": "item", "changed": true, "gid": 0, "group": "root", "item": {"atime": 1617813389.4215074, "ctime": 1617813389.188506, "dev": 64769, "gid": 0, "gr_name": "root", "inode": 21659241, "isblk": false, "ischr": false, "isdir": false, "isfifo": false, "isgid": false, "islnk": false, "isreg": true, "issock": false, "isuid": false, "mode": "0600", "mtime": 1617813389.188506, "nlink": 1, "path": "/etc/kubernetes/ssl/ca.crt", "pw_name": "root", "rgrp": false, "roth": false, "rusr": true, "size": 1066, "uid": 0, "wgrp": false, "woth": false, "wusr": true, "xgrp": false, "xoth": false, "xusr": false}, "mode": "0644", "owner": "root", "path": "/etc/kubernetes/ssl/ca.crt", "secontext": "unconfined_u:object_r:etc_t:s0", "size": 1066, "state": "file", "uid": 0}
[...]
changed: [instance-2] => (item={'uid': 0, 'woth': False, 'mtime': 1617813389.188506, 'inode': 21659242, 'isgid': False, 'size': 1078, 'roth': False, 'isuid': False, 'isreg': True, 'pw_name': 'root', 'gid': 0, 'ischr': False, 'wusr': True, 'xoth': False, 'rusr': True, 'nlink': 1, 'issock': False, 'rgrp': False, 'gr_name': 'root', 'path': '/etc/kubernetes/ssl/front-proxy-ca.crt', 'xusr': False, 'atime': 1617813389.1905057, 'isdir': False, 'ctime': 1617813389.188506, 'isblk': False, 'xgrp': False, 'dev': 64769, 'wgrp': False, 'isfifo': False, 'mode': '0600', 'islnk': False}) => {"ansible_loop_var": "item", "changed": true, "gid": 0, "group": "root", "item": {"atime": 1617813389.1905057, "ctime": 1617813389.188506, "dev": 64769, "gid": 0, "gr_name": "root", "inode": 21659242, "isblk": false, "ischr": false, "isdir": false, "isfifo": false, "isgid": false, "islnk": false, "isreg": true, "issock": false, "isuid": false, "mode": "0600", "mtime": 1617813389.188506, "nlink": 1, "path": "/etc/kubernetes/ssl/front-proxy-ca.crt", "pw_name": "root", "rgrp": false, "roth": false, "rusr": true, "size": 1078, "uid": 0, "wgrp": false, "woth": false, "wusr": true, "xgrp": false, "xoth": false, "xusr": false}, "mode": "0644", "owner": "root", "path": "/etc/kubernetes/ssl/front-proxy-ca.crt", "secontext": "unconfined_u:object_r:etc_t:s0", "size": 1078, "state": "file", "uid": 0}
```

champtar commented:
Is it only on the second control-plane ?

floryut commented:
> Is it only on the second control-plane ?

Looks like it 🤔

champtar commented on roles/kubernetes/kubeadm/tasks/main.yml:
```diff
@@ -128,12 +128,38 @@
     group: root
-    mode: "0644"
+    mode: 0600
+  with_items: "{{ pki_key_files.files }}"
```
use `with_fileglob`

champtar reviewed (changes requested):
Are the default kubeadm file permissions wrong ? or there an old kubespray workaround that mess permissions and we should also fix that ?

oomichi commented:
@floryut Could you check @champtar's comments?
